### PR TITLE
Bring together amici and pypesto petab handling

### PR DESCRIPTION
**Pull request by yannikschaelte**
use just one pipeline -- easier to handle

breaks:
* amici.plist is no longer used, i.e. potentially too many parameters are computed.
* the parameter_mapping structure passed to AmiciObjective has changed. only of effect if amici is used without petab.

### CODE REVIEW

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/objective/petab_import.py  -1
+ pypesto/objective/amici_objective.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5190639&bid=17027140)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/objective/amici_objective.py  -1
+ pypesto/objective/petab_import.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5190639&bid=17027140)

File: pypesto/objective/amici_objective.py [hunk 1]
```diff
@@ -2,13 +2,18 @@
...
+from collections import OrderedDict
+
 from .objective import Objective
 from .constants import MODE_FUN, MODE_RES, FVAL, GRAD, HESS, RES, SRES, RDATAS
+from .options import ObjectiveOptions
 
 try:
     import amici
+    import amici.petab_objective
 except ImportError:
```
Comment (dweindl) [inline pypesto/objective/amici_objective.py]: Not sure if this has any effect. If import fails, the typehints below will cause trouble. Would have to quote all amici types I think.

File: pypesto/objective/amici_objective.py [hunk 2]
```diff
@@ -289,29 +277,35 @@ def _call_amici(
...
+        par_sim_ids = list(self.amici_model.getParameterIds())
+
         for data_ix, rdata in enumerate(rdatas):
             log_simulation(data_ix, rdata)
 
             # check if the computation failed
             if rdata['status'] < 0.0:
                 return self.get_error_output(rdatas)
 
+            condition_map_sim_var = self.parameter_mapping[data_ix][2]
```
Comment (dweindl) [inline pypesto/objective/amici_objective.py]: Replace `2` by something descriptive?

File: pypesto/objective/amici_objective.py [hunk 3]
```diff
@@ -363,73 +363,16 @@ def get_error_output(self, rdatas):
...
         """
         Use the stored steadystate as well as the respective  sensitivity (
         if available) and parameter value to approximate the steadystate at
         the current parameters using a zeroth or first order taylor
         approximation:
         x_ss(x') = x_ss(x) [+ dx_ss/dx(x)*(x'-x)]
         """
-        mapping = self.mapping_par_opt_to_par_sim[condition_ix]
-        x_sim = map_par_opt_to_par_sim(mapping, self.x_ids, x)
+        mapping = self.parameter_mapping[condition_ix][2]
```
Comment (dweindl) [inline pypesto/objective/amici_objective.py]: 2 as above

File: pypesto/objective/amici_objective.py [hunk 4]
```diff
@@ -571,78 +500,74 @@ def create_plist_from_par_opt_to_par_sim(mapping_par_opt_to_par_sim):
...
-    return mapping_scale_opt_to_scale_sim
-
-
-def add_sim_grad_to_opt_grad(par_opt_ids,
-                             mapping_par_opt_to_par_sim,
-                             sim_grad,
-                             opt_grad,
-                             coefficient: float = 1.0):
+def create_identity_parameter_mapping(
+        x_ids: List[str], x_scales: List[int], n_conditions: int):
```
Comment (dweindl) [inline pypesto/objective/amici_objective.py]: return value typehint?

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/objective/petab_import.py  -1
+ pypesto/objective/amici_objective.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5190639&bid=17027140)

**Comment by yannikschaelte** [comment]
> Good to remove that redundancy.
> What's the plan regarding the broken functionality?

will make an issue for plist

File: pypesto/objective/amici_objective.py (hunk 1 above)
Comment (yannikschaelte) [reply pypesto/objective/amici_objective.py]: right. done, also for petab.

File: pypesto/objective/amici_objective.py (hunk 2 above)
Comment (yannikschaelte) [reply pypesto/objective/amici_objective.py]: that's just the index ... I know, a 6-tuple is not ideal to have. maybe we should create something more meaningful from the mapping object, make a class, ...?

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/objective/petab_import.py  -1
+ pypesto/objective/amici_objective.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5190639&bid=17027140)

File: pypesto/objective/amici_objective.py (hunk 2 above)
Comment (dweindl) [reply pypesto/objective/amici_objective.py]: Some `IDX_PAR_MAP_SIM_VAR=...` would already do the job for me.

Comment (yannikschaelte) [reply pypesto/objective/amici_objective.py]: sounds good

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/objective/amici_objective.py  -1
+ pypesto/objective/petab_import.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5190639&bid=17027140)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/objective/petab_import.py  -1
+ pypesto/objective/amici_objective.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5190639&bid=17027140)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ pypesto/objective/amici_objective.py  -1
+ pypesto/objective/petab_import.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=5190639&bid=17027140)

**Pull request merged**